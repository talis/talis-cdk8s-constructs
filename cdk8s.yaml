language: typescript
imports:
  - k8s@1.30.0
