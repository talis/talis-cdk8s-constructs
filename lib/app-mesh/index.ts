export * from "./ingress-gateway-nlb";
export * from "./app-service";
