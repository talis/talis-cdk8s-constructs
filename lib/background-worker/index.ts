export * from "./background-worker-props";
export * from "./background-worker";
