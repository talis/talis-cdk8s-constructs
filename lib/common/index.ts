export * from "./container-props";
export * from "./env-util";
export * from "./name-util";
export * from "./statefulset-util";
export * from "./value-util";
export * from "./workload-props";
