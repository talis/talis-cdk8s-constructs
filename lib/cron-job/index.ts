export * from "./cron-job-props";
export * from "./cron-job";
