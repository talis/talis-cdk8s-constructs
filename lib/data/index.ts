export * from "./config-map";
export * from "./secret";
export * from "./factory";
