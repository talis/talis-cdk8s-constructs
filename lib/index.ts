export * from "./cron-job";
export * from "./background-worker";
export * from "./data";
export * from "./job";
export * from "./redis";
export * from "./talis-chart";
export * from "./web-service";
