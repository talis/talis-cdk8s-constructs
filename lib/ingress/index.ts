export * from "./ingress-props";
export * from "./ingress";
