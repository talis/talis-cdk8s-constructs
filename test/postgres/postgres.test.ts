import { Chart, Testing } from "cdk8s";
import { KubeService, KubeStatefulSet, Quantity } from "../../imports/k8s";
import { Postgres, PostgresProps } from "../../lib";
import { makeChart } from "../test-util";

const requiredProps = {
  release: "v1",
};

function synthPostgres(props: PostgresProps = requiredProps) {
  const chart = Testing.chart();
  new Postgres(chart, "postgres-test", props);
  const results = Testing.synth(chart);
  return results;
}

describe("Postgres", () => {
  describe("Props", () => {
    test("Minimal required props", () => {
      const chart = Testing.chart();
      new Postgres(chart, "postgres-test", requiredProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();
    });

    test("All the props", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        namespace: "test",
        labels: {
          app: "my-app",
          environment: "test",
          region: "local",
        },
      });
      const selectorLabels = {
        app: "my-app",
        role: "postgres",
        instance: "test",
      };

      const allProps: Required<PostgresProps> = {
        ...requiredProps,
        selectorLabels,
        storageSize: Quantity.fromString("20Gi"),
        priorityClassName: "test",
        env: [
          {
            name: "POSTGRES_PASSWORD",
            value: "secret123",
          },
        ],
        envFrom: [
          {
            configMapRef: {
              name: "test",
            },
          },
        ],
        resources: {
          limits: {
            cpu: Quantity.fromString("100m"),
            memory: Quantity.fromString("500Mi"),
          },
        },
        initContainers: [
          {
            name: "init-container",
            image: "busybox:1.35.0",
            command: ["/bin/sh", "-c", "echo hello"],
          },
        ],
        volumes: [
          {
            name: "tmp-dir",
            emptyDir: {},
          },
        ],
        volumeMounts: [
          {
            name: "tmp-dir",
            mountPath: "/tmp",
          },
        ],
      };
      new Postgres(chart, "postgres-test", allProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();

      const statefulSet = results.find((obj) => obj.kind === "StatefulSet");
      expect(statefulSet).toHaveAllProperties(allProps, [
        "release",
        "selectorLabels",
        "storageSize",
        "resources",
      ]);
    });

    test("selectorLabels can override app", () => {
      const results = synthPostgres({
        ...requiredProps,
        selectorLabels: { app: "foobar" },
      });
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty("metadata.labels.app", "foobar");
      expect(sts).toHaveProperty("spec.selector.matchLabels.app", "foobar");
      expect(sts).toHaveProperty("spec.template.metadata.labels.app", "foobar");
    });

    test("selectorLabels can override role", () => {
      const results = synthPostgres({
        ...requiredProps,
        selectorLabels: { role: "database" },
      });
      const sts = results.find((obj) => obj.kind === "StatefulSet");
      expect(sts).toHaveProperty("metadata.labels.role", "database");
      expect(sts).toHaveProperty("spec.selector.matchLabels.role", "database");
      expect(sts).toHaveProperty(
        "spec.template.metadata.labels.role",
        "database"
      );
    });
  });

  describe("Object instances", () => {
    test("Exposes service object through property", () => {
      const chart = makeChart();
      const postgres = new Postgres(chart, "postgres-test", requiredProps);
      expect(postgres.service).toBeDefined();
      expect(postgres.service).toBeInstanceOf(KubeService);
      expect(postgres.service.name).toEqual("postgres-test");
    });

    test("Exposes statefulSet object through property", () => {
      const chart = makeChart();
      const postgres = new Postgres(chart, "postgres-test", requiredProps);
      expect(postgres.statefulSet).toBeDefined();
      expect(postgres.statefulSet).toBeInstanceOf(KubeStatefulSet);
      expect(postgres.statefulSet.name).toEqual("postgres-test-sts");
    });
  });

  describe("Container release", () => {
    test("Default container release", () => {
      const results = synthPostgres();
      const postgres = results.find((obj) => obj.kind === "StatefulSet");
      expect(postgres).toHaveProperty(
        "spec.template.spec.containers[0].image",
        "postgres:v1"
      );
    });

    test("Container release set explicitly", () => {
      const results = synthPostgres({
        ...requiredProps,
        release: "12345",
      });
      const postgres = results.find((obj) => obj.kind === "StatefulSet");
      expect(postgres).toHaveProperty(
        "spec.template.spec.containers[0].image",
        "postgres:12345"
      );
    });
  });

  describe("getDnsName", () => {
    test("Builds a DNS name for the first Pod", () => {
      const chart = makeChart();
      const postgres = new Postgres(chart, "postgres-test", requiredProps);
      expect(postgres.getDnsName()).toBe("postgres-test-sts-0.postgres-test");
    });

    const tests: [number, string][] = [
      [0, "postgres-test-sts-0.postgres-test"],
      [1, "postgres-test-sts-1.postgres-test"],
      [3, "postgres-test-sts-3.postgres-test"],
    ];
    tests.forEach(([replica, expected]) => {
      test("Builds a string from non-empty parts", () => {
        const chart = makeChart();
        const postgres = new Postgres(chart, "postgres-test", requiredProps);
        expect(postgres.getDnsName(replica)).toBe(expected);
      });
    });
  });
});
