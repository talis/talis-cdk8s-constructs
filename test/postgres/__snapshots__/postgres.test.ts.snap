// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Postgres > Props > All the props 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "labels": {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "postgres",
      },
      "name": "test-postgres-test-c8f77601",
      "namespace": "test",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 5432,
          "protocol": "TCP",
        },
      ],
      "selector": {
        "app": "my-app",
        "instance": "test",
        "role": "postgres",
      },
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": {
      "labels": {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "postgres",
      },
      "name": "test-postgres-test-postgres-test-sts-c8803176",
      "namespace": "test",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "app": "my-app",
          "instance": "test",
          "role": "postgres",
        },
      },
      "serviceName": "test-postgres-test-c8f77601",
      "template": {
        "metadata": {
          "labels": {
            "app": "my-app",
            "environment": "test",
            "instance": "test",
            "region": "local",
            "release": "v1",
            "role": "postgres",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "POSTGRES_PASSWORD",
                  "value": "secret123",
                },
              ],
              "envFrom": [
                {
                  "configMapRef": {
                    "name": "test",
                  },
                },
              ],
              "image": "postgres:v1",
              "livenessProbe": {
                "exec": {
                  "command": [
                    "pg_isready",
                  ],
                },
                "failureThreshold": 3,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "name": "postgres",
              "ports": [
                {
                  "containerPort": 5432,
                },
              ],
              "readinessProbe": {
                "exec": {
                  "command": [
                    "pg_isready",
                  ],
                },
                "failureThreshold": 3,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "resources": {
                "limits": {
                  "cpu": "100m",
                  "memory": "500Mi",
                },
              },
              "volumeMounts": [
                {
                  "mountPath": "/var/lib/postgresql/data",
                  "name": "postgres-data",
                  "subPath": "postgres",
                },
                {
                  "mountPath": "/tmp",
                  "name": "tmp-dir",
                },
              ],
            },
          ],
          "initContainers": [
            {
              "command": [
                "/bin/sh",
                "-c",
                "echo hello",
              ],
              "image": "busybox:1.35.0",
              "name": "init-container",
            },
          ],
          "priorityClassName": "test",
          "volumes": [
            {
              "emptyDir": {},
              "name": "tmp-dir",
            },
          ],
        },
      },
      "volumeClaimTemplates": [
        {
          "metadata": {
            "name": "postgres-data",
          },
          "spec": {
            "accessModes": [
              "ReadWriteOnce",
            ],
            "resources": {
              "requests": {
                "storage": "20Gi",
              },
            },
            "storageClassName": "general-purpose-delete",
          },
        },
      ],
    },
  },
]
`;

exports[`Postgres > Props > Minimal required props 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "labels": {
        "instance": "postgres-test",
        "release": "v1",
        "role": "postgres",
      },
      "name": "test-postgres-test-c8f77601",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 5432,
          "protocol": "TCP",
        },
      ],
      "selector": {
        "instance": "postgres-test",
        "role": "postgres",
      },
    },
  },
  {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": {
      "labels": {
        "instance": "postgres-test",
        "release": "v1",
        "role": "postgres",
      },
      "name": "test-postgres-test-postgres-test-sts-c8803176",
    },
    "spec": {
      "replicas": 1,
      "selector": {
        "matchLabels": {
          "instance": "postgres-test",
          "role": "postgres",
        },
      },
      "serviceName": "test-postgres-test-c8f77601",
      "template": {
        "metadata": {
          "labels": {
            "instance": "postgres-test",
            "release": "v1",
            "role": "postgres",
          },
        },
        "spec": {
          "containers": [
            {
              "image": "postgres:v1",
              "livenessProbe": {
                "exec": {
                  "command": [
                    "pg_isready",
                  ],
                },
                "failureThreshold": 3,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "name": "postgres",
              "ports": [
                {
                  "containerPort": 5432,
                },
              ],
              "readinessProbe": {
                "exec": {
                  "command": [
                    "pg_isready",
                  ],
                },
                "failureThreshold": 3,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "resources": {
                "limits": {
                  "cpu": "100m",
                  "memory": "500Mi",
                },
              },
              "volumeMounts": [
                {
                  "mountPath": "/var/lib/postgresql/data",
                  "name": "postgres-data",
                  "subPath": "postgres",
                },
              ],
            },
          ],
          "priorityClassName": "database",
        },
      },
      "volumeClaimTemplates": [
        {
          "metadata": {
            "name": "postgres-data",
          },
          "spec": {
            "accessModes": [
              "ReadWriteOnce",
            ],
            "resources": {
              "requests": {
                "storage": "20Gi",
              },
            },
            "storageClassName": "general-purpose-delete",
          },
        },
      ],
    },
  },
]
`;
